components/director/internal/domain/eventing/service.go

```diff
@@ -70,6 +80,59 @@ func (s *service) CleanupAfterUnregisteringApplication(ctx context.Context, appI
 	return model.NewEmptyApplicationEventingConfig()
 }
 
+func (s *service) DeleteAfterRemovingRuntimeFromScenarios(ctx context.Context, runtimeID, tenantID string, currentRuntimeScenarios, scenariosToRemove []string) error {
+	appLabels, err := s.scenariosService.GetApplicationScenarioLabelsForAnyMatchingScenario(ctx, scenariosToRemove)
+	if err != nil {
+		return err
+	}
+	for _, label := range appLabels {
+		applicationScenarios, err := labelpkg.ValueToStringsSlice(label.Value)
+		if err != nil {
+			return err
+		}
+		if len(str.IntersectSlice(currentRuntimeScenarios, applicationScenarios)) > 1 { // there is more than one common scenario
+			break
+		} else {
+			defaultEventingLabelKey := GetDefaultEventingForAppLabelKey(uuid.MustParse(label.ObjectID))
+			err = s.labelRepo.Delete(ctx, tenantID, model.RuntimeLabelableObject, runtimeID, defaultEventingLabelKey)
+			if err != nil && !apperrors.IsNotFoundError(err) {
+				return errors.Wrapf(err, "while deleting deefault eventing label")
+			}
+		}
+	}
+	return nil
+}
+
+func (s *service) DeleteAfterRemovingApplicationFromScenarios(ctx context.Context, applicationID, tenantID string, currentApplicationScenarios, scenariosToRemove []string) error {
+	runtime, foundDefault, err := s.GetDefaultRuntimeForAppEventing(ctx, tenantID, uuid.MustParse(applicationID))
+	if err != nil {
+		return err
+	}
+	if foundDefault {
+		defaultEventingRuntimeID := runtime.ID
+		runtimeLabels, err := s.scenariosService.GetRuntimeScenarioLabelsForAnyMatchingScenario(ctx, scenariosToRemove)
+		if err != nil {
+			return err
+		}
+		for _, runtimeLabel := range runtimeLabels {
+			if runtimeLabel.ObjectID == defaultEventingRuntimeID {
+				runtimeScenarios, err := labelpkg.ValueToStringsSlice(runtimeLabel.Value)
+				if err != nil {
+					return err
+				}
+				if len(str.IntersectSlice(currentApplicationScenarios, runtimeScenarios)) == 1 {
```

Comment:
What happens if the app and the runtime has two common scenarios and both of them are removed with one call. This means that the len here will be two, but still there aren't any scenarios left.

More robust check can be:

```go
if len(str.IntersectSlice(inputScenarios, runtimeScenarios)) == 0
```

This way you say that you remove only if with the new scenario labels of the application there are no common scenarios with the runtime anymore.

For this to work you need to pass the input in SetLabel and not the current scenarios.